### PR TITLE
feat: more arithmetic optimizations

### PR DESCRIPTION
**Pull request by charles-cooper**
### How to verify it
tests pass

### Commit Message
```
    refactor IR optimizer
    
    this is a small rewrite of the IR optimizer. it changes the structure of
    the binop optimizations so that it is easier to add more optimizations.
    it also refactors the `clamp` optimizations to be in terms of an
    `assert` statement, so that the clamp conditions can be optimized using
    the binop optimizer code.
```

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

**Review by skellet0r — APPROVED** [review]
🚀 amazing

File: vyper/ir/optimizer.py [hunk 1]
```diff
@@ -1,191 +1,389 @@
...
+    # to avoid getting clobbered by other branches)
+    # TODO: rethink structure of is_truthy optimizations.
+    elif is_truthy:
+        if binop == "eq":
+            # x == 0xff...ff => ~x == 0
+            if _int(args[1], UNSIGNED) == 2 ** 256 - 1:
+                new_val = "iszero"
+                new_args = [["not", args[0]]]
+            else:
+                # (eq x y) has the same truthyness as (iszero (sub x y))
```
Comment (fubuloubu) [inline vyper/ir/optimizer.py:219]: Overflow shouldn't affect this, right?

File: vyper/ir/optimizer.py [hunk 2]
```diff
@@ -1,191 +1,389 @@
...
+        # no rule needed for "ne" as it will get compiled to
+        # `(iszero (eq x y))` anyways.
+
+        # TODO can we do this?
+        # if val == "div":
+        #    val = "gt"
+        #    args = ["iszero", args]
+
+        elif binop in {"sgt", "gt", "slt", "lt"} and _is_int(args[1]):
+            assert unsigned == (not binop.startswith("s")), "yikes"
```
Comment (fubuloubu) [inline vyper/ir/optimizer.py:231]: better error message
```suggestion
            assert unsigned == (not binop.startswith("s")), "signed opcodes should start with s"
```

**Review by fubuloubu — APPROVED** [review]

File: vyper/ir/optimizer.py (hunk 1 above)
Comment (charles-cooper) [reply vyper/ir/optimizer.py:219]: no -- for instance, `sub 1 2` returns `-1`
Screenshot from 2022-05-09 23-37-48

**Review by fubuloubu — APPROVED** [review]

**Pull request merged**